docs(cookbook): add versioned data types tutorial (#148)
amzn/ion-schema: docs/cookbook/versioned-types.md

formats that evolve on a release schedule. Using an Ion `timestamp` with day or
month [precision](../isl-2-0/spec#timestamp_precision) preserves range support,
so you can write constraints like `valid_values: range::[2024-06T, max]` to mean
"version 2024-06 or later". The `timestamp_precision` constraint on the version
field ensures all version values have consistent precision, preventing unexpected
range comparison behavior from mixed precisions:

```ion
type::{
  name: CalVerExample,
  fields: closed::{
    version: { occurs: required, timestamp_precision: month },
    name: { occurs: required, type: string },
    color: string,  // since 2024-06
  },
  any_of: [
    { fields: { color: nothing } },
    { fields: { version: { valid_values: range::[2024-06T, max] } } },
  ],
}
```

[Semantic Versioning](https://semver.org/) (SemVer) uses a
`MAJOR.MINOR.PATCH` string to convey compatibility information in
the version number itself.

If you use a string-based version (such as SemVer), you cannot use
ranges. Instead, enumerate the valid versions explicitly with
[`valid_values`](../isl-2-0/spec#valid_values):

```ion
  any_of: [
    { fields: { color: nothing } },

>     { fields: { version: { valid_values: ["2.0", "2.1", "3.0"] } } },

Reviewer: Nit: should the example be ["2.0.0", "2.1.0", "3.0.0"]?

  ],
```

(A `regex` constraint is also possible and may be useful for semantic versions that include build metadata.)

The remainder of this article uses sequential integer versions for clarity.

## Basic example

Suppose you have a `Widget` type that started at version 1 with just a `name`
field. In version 2, a `color` field was added. In version 3, `weight` and
`dimensions` were added.

```ion
$ion_schema_2_0

type::{
  name: Widget,
  fields: closed::{
    version: { type: int, occurs: required },
    name: { type: string, occurs: required },
    color: string,       // since v2
    weight: decimal,     // since v3
    dimensions: string,  // since v3
  },
  // Optional fields added in v2:
  // Either `color` is absent, or `version` >= 2.
  any_of: [
    { fields: { color: nothing } },
    { fields: { version: { valid_values: range::[2, max] } } },
  ],
  // Optional fields added in v3:
  // Either `weight` and `dimensions` are absent, or `version` >= 3.
  any_of: [
    { fields: { weight: nothing, dimensions: nothing } },
    { fields: { version: { valid_values: range::[3, max] } } },
  ],
}
```

### How it works

> Each `any_of` encodes an implication: _if field X is present, then version must

Author: ```suggestion
Each [`any_of`](../isl-2-0/spec#any_of) encodes an implication: _if field X is present, then version must
```

be at least N._ Written in propositional logic:

> present(field) &rarr; version &ge; N

Which is equivalent to the disjunction:

> &not;present(field) &or; version &ge; N

In Ion Schema, "not present" is expressed by constraining the field to
[`nothing`](../isl-2-0/spec#built-in-types) (the empty type — no value can

> satisfy it, so the field must be absent). "Version is at least N" is expressed

Author: ```suggestion
[`nothing`](../isl-2-0/spec#built-in-types) (no value can satisfy the
empty type, so the field must be absent). "Version is at least N" is expressed
```

by constraining the `version` field with `valid_values: range::[N, max]`.

The [`any_of`](../isl-2-0/spec#any_of) constraint requires that at least one of

> its alternatives is satisfied — matching the logical "or" of the disjunction.

Author: ```suggestion
```

(For more background on expressing logical relationships in Ion Schema, see
[Expressing logical relationships between fields](logical-relationships).)

### Validation examples

The following values are **valid** for the `Widget` type:

```ion

> // v1 — only base fields

Author: ```suggestion
// v1 - only base fields
```

{ version: 1, name: "sprocket" }

> // v2 — includes color

Author: ```suggestion
// v2 - includes color
```

{ version: 2, name: "gear", color: "red" }

> // v3 — includes all fields

Author: ```suggestion
// v3 - includes all fields
```

{ version: 3, name: "bolt", color: "zinc", weight: 4.5, dimensions: "M8x30" }

// v2 without the optional v2 field (color is not required)
{ version: 2, name: "washer" }
```

The following values are **invalid**:

```ion
// INVALID: v1 data cannot have `color`
{ version: 1, name: "sprocket", color: "blue" }

// INVALID: v2 data cannot have `weight` (added in v3)
{ version: 2, name: "gear", weight: 1.2 }

// INVALID: v2 data cannot have `dimensions` (added in v3)
{ version: 2, name: "bolt", dimensions: "M6x20" }
```

## Extended example: an evolving API response

This example demonstrates several real-world version evolution scenarios in a

> single type — a field changing types, a field being removed, and an enum gaining

Author: ```suggestion
single type: a field changing types, a field being removed, and an enum gaining
```

the `fields` constraint (any non-null value). The `any_of` then narrows it:
either the field satisfies `int`, or the version is 2+ and the field satisfies
`int` or `string`. This means v1 data must have an integer `request_id`, while
v2+ data can have either `int` or `string`.

### Scenario: field removed in a later version

The `metadata` field is confined to versions 2 and 3 with a single `any_of`:
either `metadata` is absent, or the version is in `range::[2, 3]`. This covers
both introduction (not allowed before v2) and removal (not allowed after v3) in
one constraint.

### Scenario: enum gains a variant

The `priority` field uses named types (`PriorityV3`, `PriorityV4`) to define the
valid enum values for each version range. A single `any_of` handles introduction,
version-specific variants, and absence with a three-way disjunction: either
`priority` is absent (versions before it was introduced), or the version is
exactly `3` and the value matches `PriorityV3`, or the version is 4+ and the
value matches `PriorityV4`.

### Validation examples

```ion
// Valid v1
{ version: 1, request_id: 42, status: "ok", payload: {} }

> // Valid v2 — request_id is now a string, metadata present

Author: ```suggestion
// Valid v2 - request_id is now a string, metadata present
```

{ version: 2, request_id: "abc-123", status: "ok",
  payload: {}, metadata: { trace_id: "x" } }

> // Valid v3 — priority with v3 variant

Author: ```suggestion
// Valid v3 - priority with v3 variant
```

{ version: 3, request_id: "def-456", status: "ok",
  payload: {}, metadata: { trace_id: "y" }, priority: high }

> // Valid v4 — metadata gone, priority uses new variant

Author: ```suggestion
// Valid v4 - metadata gone, priority uses new variant
```

> version &ge; N &rarr; present(field)

Which is equivalent to:

> version &lt; N &or; present(field)

```ion
$ion_schema_2_0

type::{
  name: WidgetStrict,
  fields: closed::{
    version: { occurs: required, type: int },
    name: { occurs: required, type: string },
    color: string,  // required since v2
  },
  // `color` not allowed in v1, required in v2+
  any_of: [
    { fields: { color: nothing, version: { valid_values: [1] } } },
    { fields: { color: { occurs: required }, version: { valid_values: range::[2, max] } } },
  ],
}
```

With this definition, a version-2 `WidgetStrict` _must_ include `color`, but a
version-1 `WidgetStrict` must not.

Note that `color` is declared as optional in `fields` (no `occurs: required`).
The conditional requirement is enforced entirely by the `any_of`. If you used
`occurs: required` on the field in `fields`, it would be required for _all_
versions, defeating the purpose.

### Three-state lifecycle: absent, optional, required

A field may pass through three states across versions: absent (not yet
introduced), optional (allowed but not required), and required. This requires

> two `any_of` constraints — one for each transition:

Author: ```suggestion
one `any_of` constraint for each transition:
```

cart and its items are versioned independently:

```ion
$ion_schema_2_0

type::{
  name: CartItem,
  fields: closed::{
    version: { occurs: required, type: int },
    product_id: { occurs: required, type: string },
    quantity: { occurs: required, type: int },
    unit_price: decimal,    // since v2
    discount: decimal,      // since v3
  },
  any_of: [
    { fields: { unit_price: nothing } },
    { fields: { version: { valid_values: range::[2, max] } } },
  ],
  any_of: [
    { fields: { discount: nothing } },
    { fields: { version: { valid_values: range::[3, max] } } },
  ],
}

type::{
  name: ShoppingCart,
  fields: closed::{
    version: { occurs: required, type: int },
    cart_id: { occurs: required, type: string },
    items: { occurs: required, type: list, element: CartItem },
    coupon_code: string,    // since v2
    shipping_estimate: {    // since v2
      timestamp_precision: day,
    },
  },
  any_of: [
    { fields: { coupon_code: nothing, shipping_estimate: nothing } },
    { fields: { version: { valid_values: range::[2, max] } } },
  ],
}
```

Each `CartItem` in the `items` list is validated against its own `version`

> field. A v1 `ShoppingCart` can contain v3 `CartItem`s and vice versa — the

Author: ```suggestion
field. A v1 `ShoppingCart` can contain v3 `CartItem`s and vice versa. The
```

{
  version: 1,
  cart_id: "cart-001",
  items: [
    { version: 1, product_id: "SKU-A", quantity: 2 },
    { version: 2, product_id: "SKU-B", quantity: 1, unit_price: 9.99 },
  ],
}

// A v2 cart with v3 items
{
  version: 2,
  cart_id: "cart-002",
  coupon_code: "SAVE10",
  shipping_estimate: 2025-03-15T,
  items: [
    { version: 3, product_id: "SKU-C", quantity: 1,
      unit_price: 24.99, discount: 2.50 },
  ],
}

// INVALID: v1 cart cannot have coupon_code
{
  version: 1,
  cart_id: "cart-003",
  coupon_code: "OOPS",
  items: [],
}

// INVALID: v1 item cannot have unit_price
{
  version: 2,
  cart_id: "cart-004",
  items: [
    { version: 1, product_id: "SKU-D", quantity: 1, unit_price: 5.00 },
  ],
}
```

## Multiple fields per version

When several fields are introduced in the same version, group them in a single
`any_of` constraint. All fields in the group should appear together in the
`nothing` alternative:

```ion
  // Fields added in v3
  any_of: [
    { fields: { weight: nothing, dimensions: nothing } },
    { fields: { version: { valid_values: range::[3, max] } } },
  ],
```

This means that _any_ of those fields being present requires version 3 or

> higher — but they do not need to co-occur. In v3+, any combination of the

Author: ```suggestion
higher, but they do not need to co-occur. In v3+, any combination of the
```

grouped fields is valid. Use separate `any_of` constraints when fields are
introduced in different versions.

## Tips

- **Closed fields are important.** Without [`closed::`](../isl-2-0/spec#fields),
  unknown fields are silently allowed regardless of version, which defeats the
  purpose of version gating.
- **Use named types for complex field constraints.** When a field's valid values
  change across versions (like the `priority` enum above), define a named type
  for each version's constraint set. This keeps the main type readable.
- **Validation cost scales linearly.** Each `any_of` constraint is evaluated
  independently during validation. For types with many version gates, this means
  validation cost grows with the number of gates. (See

>   [`ion-schema-rust#253`](https://github.com/amazon-ion/ion-schema-rust/issues/253)`)

Reviewer: nit: double ) at the end of this line.
